File: backend/app/api/v1/ai_usage.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_admin
from app.core.database import get_db
from app.core.enums import AiUsageConflictStatus, AiUsageFinalizationStatus, AiUsageMappingResolutionStatus, AiUsagePricingResolutionStatus
from app.core.idempotency import require_idempotency_key
from app.models.admin import Admin
from app.schemas.ai_usage import AIUsageBatchResolveMappingsRequest, AIUsageBatchResolvePricingRequest, AIUsageConflictReviewRequest, AIUsageResolvePricingRequest
from app.services.ai_usage_service import (
    UsageFilters,
    get_conflict_detail,
    get_usage,
    list_usage,
    mark_conflict_reviewed,
    resolve_missing_mappings,
    resolve_missing_pricing,
    resolve_usage_pricing,
    summarize_usage,
)

router = APIRouter(prefix="/ai/usage", tags=["ai-usage"])


def _filters(
    product_deployment_id: UUID | None,
    organization_id: UUID | None,
    product_organization_id: str | None,
    provider: str | None,
    product_model_id: str | None,
    usage_from: datetime | None,
    usage_to: datetime | None,
    pricing_catalog_id: UUID | None = None,
    pricing_version_id: UUID | None = None,
    cost_currency: str | None = None,
    pricing_resolution_status: AiUsagePricingResolutionStatus | None = None,
    mapping_resolution_status: AiUsageMappingResolutionStatus | None = None,
    conflict_status: AiUsageConflictStatus | None = None,
    finalization_status: AiUsageFinalizationStatus | None = None,
    product_usage_id: str | None = None,
) -> UsageFilters:
    return UsageFilters(
        product_deployment_id=product_deployment_id,
        organization_id=organization_id,
        product_organization_id=product_organization_id,
        provider=provider,
        product_model_id=product_model_id,
        usage_from=usage_from,
        usage_to=usage_to,
        pricing_catalog_id=pricing_catalog_id,
        pricing_version_id=pricing_version_id,
        cost_currency=cost_currency,
        pricing_resolution_status=pricing_resolution_status,
        mapping_resolution_status=mapping_resolution_status,
        conflict_status=conflict_status,
        finalization_status=finalization_status,
        product_usage_id=product_usage_id,
    )


@router.get("")
async def ai_usage_index(
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
    limit: int = Query(default=25, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    product_deployment_id: UUID | None = None,
    organization_id: UUID | None = None,
    product_organization_id: str | None = None,
    provider: str | None = None,
    product_model_id: str | None = None,
    usage_from: datetime | None = None,
    usage_to: datetime | None = None,
    pricing_catalog_id: UUID | None = None,
    pricing_version_id: UUID | None = None,
    cost_currency: str | None = None,
    pricing_resolution_status: AiUsagePricingResolutionStatus | None = None,
    mapping_resolution_status: AiUsageMappingResolutionStatus | None = None,
    conflict_status: AiUsageConflictStatus | None = None,
    finalization_status: AiUsageFinalizationStatus | None = None,
    product_usage_id: str | None = None,
) -> dict:
    payload = list_usage(
        db,
        _filters(product_deployment_id, organization_id, product_organization_id, provider, product_model_id, usage_from, usage_to, pricing_catalog_id, pricing_version_id, cost_currency, pricing_resolution_status, mapping_resolution_status, conflict_status, finalization_status, product_usage_id),
        limit=limit,
        offset=offset,
    )
    return {"success": True, "data": payload.model_dump(mode="json")}


@router.get("/summary")
async def ai_usage_summary(
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
    product_deployment_id: UUID | None = None,
    organization_id: UUID | None = None,
    product_organization_id: str | None = None,
    provider: str | None = None,
    product_model_id: str | None = None,
    usage_from: datetime | None = None,
    usage_to: datetime | None = None,
) -> dict:
    payload = summarize_usage(db, _filters(product_deployment_id, organization_id, product_organization_id, provider, product_model_id, usage_from, usage_to))
    return {"success": True, "data": payload.model_dump(mode="json")}


@router.post("/resolve-missing-pricing")
async def ai_usage_resolve_missing_pricing(
    payload: AIUsageBatchResolvePricingRequest,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
    idempotency_key: str = Depends(require_idempotency_key),
) -> dict:
    return {"success": True, "data": resolve_missing_pricing(db, payload, idempotency_key, current_admin)}


@router.post("/resolve-mappings")
async def ai_usage_resolve_mappings(
    payload: AIUsageBatchResolveMappingsRequest,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
    idempotency_key: str = Depends(require_idempotency_key),
) -> dict:
    return {"success": True, "data": resolve_missing_mappings(db, payload, idempotency_key, current_admin)}


@router.get("/{usage_id}")
async def ai_usage_detail(
    usage_id: UUID,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
) -> dict:
    return {"success": True, "data": get_usage(db, usage_id).model_dump(mode="json")}


@router.post("/{usage_id}/resolve-pricing")
async def ai_usage_resolve_pricing(
    usage_id: UUID,
    payload: AIUsageResolvePricingRequest,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
    idempotency_key: str = Depends(require_idempotency_key),
) -> dict:
    return {"success": True, "data": resolve_usage_pricing(db, usage_id, payload, idempotency_key, current_admin)}


@router.get("/{usage_id}/conflict")
async def ai_usage_conflict_detail(
    usage_id: UUID,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
) -> dict:
    return {"success": True, "data": get_conflict_detail(db, usage_id).model_dump(mode="json")}


@router.post("/{usage_id}/conflict/mark-reviewed")
async def ai_usage_conflict_mark_reviewed(
    usage_id: UUID,
    payload: AIUsageConflictReviewRequest,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin),
    idempotency_key: str = Depends(require_idempotency_key),
) -> dict:
    return {"success": True, "data": mark_conflict_reviewed(db, usage_id, payload, idempotency_key, current_admin)}

File: frontend/src/lib/aiUsage.ts
import { apiRequest } from "@/lib/api";
import type {
  AiUsageBatchResolutionResponse,
  AiUsageConflictDetail,
  AiUsageListPayload,
  AiUsageRecord,
  AiUsageSummary,
  AiUsageSyncRun,
  AiUsageSyncRunListPayload,
  AiUsageSyncState,
  ApiResponse
} from "@/lib/types";

export type AiUsageFilters = {
  limit?: number;
  offset?: number;
  product_deployment_id?: string;
  organization_id?: string;
  product_organization_id?: string;
  provider?: string;
  product_model_id?: string;
  usage_from?: string;
  usage_to?: string;
  pricing_catalog_id?: string;
  pricing_version_id?: string;
  cost_currency?: string;
  pricing_resolution_status?: string;
  mapping_resolution_status?: string;
  conflict_status?: string;
  finalization_status?: string;
  product_usage_id?: string;
};

function queryString(filters: AiUsageFilters = {}) {
  const params = new URLSearchParams();
  Object.entries(filters).forEach(([key, value]) => {
    if (value !== undefined && value !== "") params.set(key, String(value));
  });
  const query = params.toString();
  return query ? `?${query}` : "";
}

export function listAiUsage(filters: AiUsageFilters = {}) {
  return apiRequest<ApiResponse<AiUsageListPayload>>(`/ai/usage${queryString(filters)}`);
}

export function summarizeAiUsage(filters: AiUsageFilters = {}) {
  return apiRequest<ApiResponse<AiUsageSummary>>(`/ai/usage/summary${queryString(filters)}`);
}

export function resolveAiUsagePricing(usageId: string, payload: { reason: string; pricing_version_id?: string | null }, idempotencyKey: string) {
  return apiRequest<ApiResponse<{ outcome: string; message: string | null; usage: AiUsageRecord | null }>>(`/ai/usage/${usageId}/resolve-pricing`, {
    method: "POST",
    headers: { "Idempotency-Key": idempotencyKey },
    json: payload
  });
}

export function resolveMissingPricing(payload: { reason: string; product_deployment_id?: string | null; limit?: number }, idempotencyKey: string) {
  return apiRequest<ApiResponse<AiUsageBatchResolutionResponse>>("/ai/usage/resolve-missing-pricing", {
    method: "POST",
    headers: { "Idempotency-Key": idempotencyKey },
    json: payload
  });
}

export function resolveMappings(payload: { reason: string; product_deployment_id?: string | null; limit?: number }, idempotencyKey: string) {
  return apiRequest<ApiResponse<AiUsageBatchResolutionResponse>>("/ai/usage/resolve-mappings", {
    method: "POST",
    headers: { "Idempotency-Key": idempotencyKey },
    json: payload
  });
}

export function getAiUsageConflict(usageId: string) {
  return apiRequest<ApiResponse<AiUsageConflictDetail>>(`/ai/usage/${usageId}/conflict`);
}

export function markAiUsageConflictReviewed(usageId: string, payload: { reason: string }, idempotencyKey: string) {
  return apiRequest<ApiResponse<AiUsageConflictDetail>>(`/ai/usage/${usageId}/conflict/mark-reviewed`, {
    method: "POST",
    headers: { "Idempotency-Key": idempotencyKey },
    json: payload
  });
}

export function syncProductAiUsage(productId: string, payload: { reason: string; limit?: number; max_pages?: number }, idempotencyKey: string) {
  return apiRequest<ApiResponse<AiUsageSyncRun>>(`/products/${productId}/sync/token-usage`, {
    method: "POST",
    headers: { "Idempotency-Key": idempotencyKey },
    json: payload
  });
}

export function getProductAiUsageSyncState(productId: string) {
  return apiRequest<ApiResponse<AiUsageSyncState | null>>(`/products/${productId}/ai-usage-sync-state`);
}

export function listProductAiUsageSyncRuns(productId: string, filters: { limit?: number; offset?: number } = {}) {
  return apiRequest<ApiResponse<AiUsageSyncRunListPayload>>(`/products/${productId}/ai-usage-sync-runs${queryString(filters)}`);
}
